### PR TITLE
currency: Make pairs.Add variadic

### PR DESCRIPTION
**Pull request by cranktakular**
# PR Description

Alters the Add method on currency.Pairs to accept an arbitrary amount of Pair objects as input, to enable mass-scale safe adding of pairs, in a way that hopefully doesn't break code that currently uses this function.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run
- [x] Test X

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: currency/pairs_test.go [hunk 1]
```diff
@@ -267,19 +267,21 @@ func TestAdd(t *testing.T) {
...
-
 	// Now test adding a pair which already exists
 	pairs = pairs.Add(p)
 	if len(pairs) != 4 {
 		t.Error("TestAdd unexpected result")
 	}
+	pairs = pairs.Add(NewPair(BTC, LTC), NewPair(ETH, USD))
+	if len(pairs) != 6 {
+		t.Error("TestAdd unexpected result")
+	}
```
Comment (shazbert) [inline currency/pairs_test.go:284]: There's a non zero chance of append not allocating a separate slice and you get weird effects and spooky ghosts 👻. Here's a test for it:

```diff
diff --git a/currency/pairs_test.go b/currency/pairs_test.go
index 33d3fa411..1eaa494f6 100644
--- a/currency/pairs_test.go
+++ b/currency/pairs_test.go
@@ -3,6 +3,7 @@ package currency
 import (
        "encoding/json"
        "errors"
+       "fmt"
        "testing"

        "github.com/stretchr/testify/assert"
@@ -282,6 +283,43 @@ func TestAdd(t *testing.T) {
        if len(pairs) != 6 {
                t.Error("TestAdd unexpected result")
        }
+
+       pairsWithExtraBaggage := make(Pairs, 0, len(pairs)+3)
+       pairsWithExtraBaggage = append(pairsWithExtraBaggage, pairs...)
+
+       brain := NewPair(BRAIN, USD)
+       withBrain := pairsWithExtraBaggage.Add(NewPair(BTC, LTC), brain)
+       if len(pairs) != 6 {
+               t.Error("TestAdd unexpected result")
+       }
+
+       assert.Equal(t, brain, withBrain[len(withBrain)-1])
+
+       badger := NewPair(BADGER, USD)
+       withBadger := pairsWithExtraBaggage.Add(NewPair(BTC, LTC), badger)
+       if len(pairs) != 6 {
+               t.Error("TestAdd unexpected result")
+       }
+
+       assert.Equal(t, badger, withBadger[len(withBadger)-1])
+
+       babyDoge := NewPair(BABYDOGE, USD)
+       withBabyDoge := pairsWithExtraBaggage.Add(NewPair(BTC, LTC), babyDoge)
+       if len(pairs) != 6 {
+               t.Error("TestAdd unexpected result")
+       }
+
+       assert.Equal(t, babyDoge, withBabyDoge[len(withBabyDoge)-1])
+
+       // check bleed back
+
+       assert.Equal(t, brain, withBrain[len(withBrain)-1])
+       assert.Equal(t, badger, withBadger[len(withBadger)-1])
+
+       fmt.Println(withBadger)
+       fmt.Println(withBrain)
+       fmt.Println(withBabyDoge)
+       fmt.Println(pairsWithExtraBaggage)
 }
```

slices.Clone call before or make(Pairs, 0, len(p) + len(Pairs) then copy original acrost.

**Review by shazbert — CHANGES_REQUESTED** [review]
This is much betterer! Thank you 🫡

**Comment by cranktakular** [comment]
Implementation overhaul risks breaking uses of this function, but as far as I can tell I didn't cause any new issues.

File: currency/pairs.go [hunk 2]
```diff
@@ -211,12 +211,17 @@ func (p Pairs) Remove(pair Pair) (Pairs, error) {
 }
 
 // Add adds a specified pair to the list of pairs if it doesn't exist
-func (p Pairs) Add(pair Pair) Pairs {
-	if p.Contains(pair, true) {
-		return p
+func (p Pairs) Add(pairs ...Pair) Pairs {
+	p2 := make(Pairs, len(pairs))
```
Comment (shazbert) [inline currency/pairs.go:215]: suggestion: Can consolidate your idea a little bit more. Let me know if this works.

```go
merge := append(slices.Clone(p), pairs...)
var filterInt int
for x := len(p); x < len(merge); x++ {
	if !merge[:len(p)+filterInt].Contains(merge[x], true) {
		merge[len(p)+filterInt] = merge[x]
		filterInt++
	}
}
return merge[:len(p)+filterInt]
```

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for improving things well done, added a suggestion.

File: exchanges/sharedtestvalues/sharedtestvalues.go [hunk 3]
```diff
@@ -168,10 +168,8 @@ func SetupCurrencyPairsForExchangeAsset(t *testing.T, exch exchange.IBotExchange
 		t.Fatal(err)
 	}
 	epLen := len(enabledPairs)
-	for i := range cp {
-		availPairs = availPairs.Add(cp[i])
-		enabledPairs = enabledPairs.Add(cp[i])
-	}
+	availPairs.Add(cp...)
+	enabledPairs.Add(cp...)
```
Comment (shazbert) [inline exchanges/sharedtestvalues/sharedtestvalues.go:172]:
```go
availPairs = availPairs.Add(cp...)
enabledPairs = enabledPairs.Add(cp...)
```

File: exchanges/subscription/subscription.go [hunk 4]
```diff
@@ -156,8 +156,6 @@ func (s *Subscription) AddPairs(pairs ...currency.Pair) {
 		return
 	}
 	s.m.Lock()
-	for _, p := range pairs {
-		s.Pairs = s.Pairs.Add(p)
-	}
+	s.Pairs.Add(pairs...)
```
Comment (shazbert) [inline exchanges/subscription/subscription.go:159]: s.Pairs = s.Pairs.Add(pairs...)

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for the changes! Just some nits on the most recent commit.

**Review by shazbert — CHANGES_REQUESTED** [review]
One thing left over then looks good to go.

```diff
diff --git a/backtester/funding/trackingcurrencies/trackingcurrencies_test.go b/backtester/funding/trackingcurrencies/trackingcurrencies_test.go
index 751089154..b3f187562 100644
--- a/backtester/funding/trackingcurrencies/trackingcurrencies_test.go
+++ b/backtester/funding/trackingcurrencies/trackingcurrencies_test.go
@@ -53,12 +53,8 @@ func TestCreateUSDTrackingPairs(t *testing.T) {
        cp3 := currency.NewPair(currency.LTC, currency.BTC)
        exchB := exch.GetBase()
        eba := exchB.CurrencyPairs.Pairs[a]
-       eba.Available = eba.Available.Add(cp)
-       eba.Enabled = eba.Enabled.Add(cp)
-       eba.Available = eba.Available.Add(cp2)
-       eba.Enabled = eba.Enabled.Add(cp2)
-       eba.Available = eba.Available.Add(cp3)
-       eba.Enabled = eba.Enabled.Add(cp3)
+       eba.Available = eba.Available.Add(cp, cp2, cp3)
+       eba.Enabled = eba.Enabled.Add(cp, cp2, cp3)
        eba.AssetEnabled = convert.BoolPtr(true)

        err = em.Add(exch)
```

**Review by shazbert — APPROVED** [review]
Thank you sir, well done!

**Review by gloriousCode — APPROVED** [review]
`Pairs.Add(t, ACK)`

**Review by thrasher- — APPROVED** [review]
Nice work @cranktakular !

**Pull request merged**